### ISSUE TITLE
[BUG] High CPU / Latency when adding/removing large number of routes

### ISSUE THREAD

**Issue by mabicca** [Azure/WALinuxAgent#2118]
Hi Everyone,

We are working on a reported behavior where a Linux VM is being used as a gateway and routes are being added/removed dynamically and in some cases, large numbers of routes at a time.

It was brought to our attention that when that happens, the waagent will spike on CPU usage and also cause a little bit of latency for a while, which sometimes last minutes until it goes back to a more stable scenario due to it processing all the routes being added.

We noticed using pings and mtr that the latency on network usually was averaging 68ms between 2 VM's on different peered vnets and it went up to 117ms.

Currently we are testing by using a script which allows us to add/remove large numbers of routes to make it easier:

````
import os
import sys
from netaddr import IPNetwork

action = sys.argv[1]
num = int(sys.argv[2])
network = IPNetwork('10.1.8.0/16')
for i, cidr in enumerate(network):
    cmd = 'ip route {} {} dev eth0 src 10.0.2.4 metric 100'.format(action, cidr)
    os.system(cmd)
    if i > num:
        print('Num of routes {}'.format(num))
        break
`````
Where you can use:

`````
python routes.py add 1000
or
python routes.py del 1000
`````
That would either add or remove the number of routes specified in the command.

This can be reproduced easily on any distribution, for our case, it was tested with Ubuntu 18.04 LTS and the latest waagent and kernel:

````
WALinuxAgent-2.2.45 running on ubuntu 18.04
Python: 3.6.9
Goal state agent: 2.2.53

Both VM's are at kernel version: 
5.4.0-1031-azure
````

It does look that this is being caused by the waagent having to process the amount of routes being added. The main ask here is to see if there is any way to make that less aggressive or optimize the logic so it doesn't introduce the latency observed. As far as CPU usage, we can see if dipping in and out everytime we restart the waagent since it will process that again, for example, this is a snapshot of mpstat while restart the waagent:

![image](https://user-images.githubusercontent.com/16481759/102543909-ddfc2d80-4068-11eb-914d-0e0da4fc3463.png)

This is the output from MTR after running it for a while. The high latency was observed for a while and that is the current problem with the report since it can affect the connectivity in these cases where routes are being added/removed dynamically. The issue is also not present with the waagent stopped (since there is no process on the routes).

![image](https://user-images.githubusercontent.com/16481759/102544133-316e7b80-4069-11eb-82fe-5d061fb07526.png)

Another point that might be interesting to explore, would be to possibly pin our process to one CPU only too, not sure what that would buy but at least other cores would be still free. This behavior is also noticed on larger VM's, in my test I used Standard DS2_V3 VMs.

Thank you,
-Marco

**Comment by mabicca** [comment]
And just to clarify, yes, the behavior is more severe with VM's that have less cores (single or dual cores).

**Comment by larohra** [comment]
@mabicca thanks for bringing this up. The Linux Agent does a periodic operation every minute to check if the network configurations changed in the VM and based on that reports the changes. Here's the snippet - 

```python
    def _operation_impl(self):
        raw_route_list = self.osutil.read_route_table()
        digest = hash_strings(raw_route_list)
        if digest != self.last_route_table_hash:
            self.last_route_table_hash = digest
            route_list = self.osutil.get_list_of_routes(raw_route_list)
            logger.info("Route table: [{0}]".format(",".join(map(networkutil.RouteEntry.to_json, route_list))))

        nic_state = self.osutil.get_nic_state()
        if nic_state != self.last_nic_state:
            description = "Initial" if self.last_nic_state == {} else "Updated"
            logger.info("{0} NIC state: [{1}]".format(description, ", ".join(map(str, nic_state.values()))))
            self.last_nic_state = nic_state
``` 

This so far seems to be the biggest contender that could be causing the latency (my guess is that due to the long list of routes, calculating the hashes is causing the CPU spikes). 

Would it be possible to get the `/var/log/waagent.log` file from the affected machine? I can check out a couple of logs to confirm if this is indeed the one causing the latency.  

Thanks!

**Comment by mabicca** [comment]
Hi @larohra , thanks for looking into this. Yes, that is definitely the reason. I can see the waagent.log going crazy when adding/removing the routes :) I can probably reproduce this again later, but if you want to run the script to add/remove routes, it is super easy to reproduce it too. 

Just let me know if you want to reproduce it, otherwise I'll spin up a VM and do it later today.

**Comment by larohra** [comment]
@mabicca thanks for confirming, I think I'll just try to reproduce it on my own VM and see how we can optimize it. I'll definitely let you know if I need more help/info regarding this.

Thanks!

**Comment by mabicca** [comment]
Hi @larohra , of course, thank you. Yes, please, let me know if you need anything else and I'll gladly help.

Thanks again!
-Marco

**Comment by mabicca** [comment]
Hi @larohra , please, let me know if you guys have any updates around this or need any further help.

Thank you,
-Marco

**Comment by ZhidongPeng** [comment]
@mabicca, I will follow up the history and get back.

**Comment by mabicca** [comment]
Thank you @ZhidongPeng !

**Comment by aarsan** [comment]
Is this:

https://github.com/Azure/WALinuxAgent/pull/2156 

the PR that is supposed to resolve the issue?

**Comment by narrieta** [comment]
@aarsan Yes, that is the fix

**Comment by narrieta** [comment]
Fixed by #2156
